src/Workspaces/Core/Portable/Workspace/Solution/ProjectOrDocumentId.cs

```diff
@@ -0,0 +1,57 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+#nullable enable
+
+using System;
+using System.Collections.Generic;
+
+namespace Microsoft.CodeAnalysis
+{
+    internal readonly struct ProjectOrDocumentId : IEquatable<ProjectOrDocumentId>
+    {
+        private readonly object? _projectOrDocumentId;
+
+        public ProjectOrDocumentId(ProjectId projectId)
+        {
+            _projectOrDocumentId = projectId ?? throw new ArgumentNullException(nameof(projectId));
+        }
+
+        public ProjectOrDocumentId(DocumentId documentId)
+        {
+            _projectOrDocumentId = documentId ?? throw new ArgumentNullException(nameof(documentId));
+        }
+
+        public bool IsDefault => _projectOrDocumentId is null;
+
+        public ProjectId ProjectId => DocumentId?.ProjectId ?? _projectOrDocumentId as ProjectId ?? throw new InvalidOperationException();
```

Comment:
>_projectOrDocumentId as ProjectId [](start = 63, length = 33)

I'd just cast `(ProjectId)_projectOrDocumentId`.

Reply:
➡️ The current form falls out from https://github.com/dotnet/roslyn/pull/40670#discussion_r362592157.